### PR TITLE
feat: Upstream `elasticsearch` component

### PR DESCRIPTION
**Pull request by nitrocode**
## what
* Upstream `elasticsearch` component
* Bumped version from 0.30.0 to 0.33.0 (latest non pre-release)

## why
* Install kibana with a security group, iam role, and configured elasticsearch

## references
* https://github.com/cloudposse/terraform-aws-elasticsearch

## notes

To pass pre-commit hook

```shell
tfenv install 0.13.5
tfenv use 0.13.5
make readme
pre-commit run --show-diff-on-failure --color=always --all-files
git add modules/elasticsearch
```

### CODE REVIEW

**Review by osterman — DISMISSED** [review]

File: modules/elasticsearch/main.tf [hunk 1]
```diff
@@ -0,0 +1,110 @@
+locals {
+  enabled = module.this.enabled
+
+  dns_zone_id                   = module.dns_delegated.outputs.default_dns_zone_id
+  vpc_id                        = module.vpc.outputs.vpc_id
+  vpc_default_security_group    = module.vpc.outputs.vpc_default_security_group_id
+  vpc_private_subnet_ids        = module.vpc.outputs.private_subnet_ids
+  elasticsearch_domain_endpoint = format("/%s/%s/%s", "elasticsearch", module.this.name, "elasticsearch_domain_endpoint")
+  elasticsearch_kibana_endpoint = format("/%s/%s/%s", "elasticsearch", module.this.name, "elasticsearch_kibana_endpoint")
+  elasticsearch_admin_password  = format("/%s/%s/%s", "elasticsearch", module.this.name, "password")
```
Comment (korenyoni) [inline modules/elasticsearch/main.tf:10]:
```suggestion
  elasticsearch_endpoint_format = "/elasticsearch/${module.this.name}/%s"
  elasticsearch_domain_endpoint = format(local.elasticsearch_endpoint_format, "elasticsearch_domain_endpoint")
  elasticsearch_kibana_endpoint = format(local.elasticsearch_endpoint_format, "elasticsearch_kibana_endpoint")
  elasticsearch_admin_password  = format(local.elasticsearch_endpoint_format, "password")
```

Just a nitpick

File: modules/elasticsearch/main.tf [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+  security_groups                = [local.vpc_default_security_group]
+  vpc_id                         = local.vpc_id
+  subnet_ids                     = local.vpc_private_subnet_ids
+  zone_awareness_enabled         = length(local.vpc_private_subnet_ids) > 1 ? true : false
+  elasticsearch_version          = var.elasticsearch_version
+  instance_type                  = var.instance_type
+  instance_count                 = length(local.vpc_private_subnet_ids)
+  availability_zone_count        = length(local.vpc_private_subnet_ids)
+  encrypt_at_rest_enabled        = true //var.encrypt_at_rest_enabled
```
Comment (korenyoni) [inline modules/elasticsearch/main.tf:30]:
```suggestion
  encrypt_at_rest_enabled        = var.encrypt_at_rest_enabled
```

`var.encrypt_at_rest_enabled` Is set in the defaults and also is mentioned in the usage documentation. It should not be ignored.

File: modules/elasticsearch/variables.tf [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+variable "kibana_subdomain_name" {
+  type        = string
+  description = "The name of the subdomain for Kibana in the DNS zone (_e.g._ `kibana`, `ui`, `ui-es`, `search-ui`, `kibana.elasticsearch`)"
+}
+
+variable "create_iam_service_linked_role" {
+  type        = bool
+  description = "Whether to create `AWSServiceRoleForAmazonElasticsearchService` service-linked role. Set it to `false` if you already have an ElasticSearch cluster created in the AWS account and AWSServiceRoleForAmazonElasticsearchService already exists. See https://github.com/terraform-providers/terraform-provider-aws/issues/5218 for more info"
```
Comment (korenyoni) [inline modules/elasticsearch/variables.tf:38]: This is a long description and should use heredoc.

```suggestion
  description = <<-EOT
  Whether to create `AWSServiceRoleForAmazonElasticsearchService` service-linked role.
  Set this to `false` if you already have an ElasticSearch cluster created in the AWS account and `AWSServiceRoleForAmazonElasticsearchService` already exists.
  See https://github.com/terraform-providers/terraform-provider-aws/issues/5218 for more information.
  EOT
```

File: modules/elasticsearch/versions.tf [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    }
+    template = {
+      source  = "hashicorp/template"
+      version = ">= 2.0"
+    }
+    null = {
+      source  = "hashicorp/null"
+      version = ">= 2.0"
+    }
+  }
```
Comment (korenyoni) [inline modules/elasticsearch/versions.tf:17]:
```suggestion
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = ">= 3.8"
    }
  }
```

I didn't run `tflint` on this component, but I'm fairly sure only the `aws` provider is used.

**Review by korenyoni — CHANGES_REQUESTED** [review]
Requested some changes. The one comment I mention is a nitpick is not a required change.

**Review by korenyoni — APPROVED** [review]
Looks good to me.

**Pull request merged**